sensor/common/context/context_provider.go

```diff
@@ -0,0 +1,77 @@
+package contextprovider
+
+import (
+	"context"
+
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/sensor/common"
+	"github.com/stackrox/rox/sensor/common/message"
+)
+
+// ContextProvider Sensor component that provides a common context.
+// This context will be cancelled and reset if Sensor enters Offline mode.
+type ContextProvider interface {
+	common.SensorComponent
+	GetContext() context.Context
+}
+
+var _ ContextProvider = (*contextProviderImpl)(nil)
+
+// NewContextProvider Initializes a new ContextProvider
+func NewContextProvider() ContextProvider {
+	return &contextProviderImpl{
+		centralReachable: concurrency.NewSignal(),
+		stopper:          concurrency.NewStopper(),
+	}
+}
+
+type contextProviderImpl struct {
+	cancelContextFn  func()
+	sensorContext    context.Context
+	centralReachable concurrency.Signal
+	stopper          concurrency.Stopper
+}
+
+// GetContext returns the sensor context. This call will block until central is reachable.
+// This blocking in behavior is needed to ensure unset or old contexts are not passed to other component.
```

Comment:
I think there are some potential problems in making this call blocking. There are couple of scenarios I can think of where this could generate incorrect results (or unnecessary processing, at the very least):

1) Imagine `deployment-a` event is received at the Kubernetes listener. Right before the handler assigns a context to the component message ([here](https://github.com/stackrox/stackrox/blob/master/sensor/kubernetes/listener/resource_event_handler_impl.go#L110)) the connection stops. A call to `GetContext()` at that line will block until central is reachable again. If, while sensor is down, `deployment-a` gets deleted, the goroutine will resume the processing of `deployment-a` event by appending a fresh (!) context (once this select block unblocks). This will lead to Sensor incorrectly sending `deployment-a` to Central during the reconciliation.

2) Imagine a message from Central arrives that all deployments need to be reprocessed. [Here](https://github.com/stackrox/stackrox/blob/master/sensor/kubernetes/eventpipeline/pipeline_impl.go#L163), before sending the event, we'd add a call to `GetContext()`. Again, if the connection breaks right before making this call, Sensor will process all deployments twice during the reconciliation. One from the Kubernetes flush, and the second when `GetContext()` unblocks. 

IMO these are two good reasons why this call shouldn't block. I think the ideal behavior is to (in both cases) return the expired context. If we _just_ switched into offline mode, and some goroutine is still processing incoming events, context provider should return a context that is expired, so this message will be stopped somewhere in the pipeline. 

Additionally, every place where `GetContext()` is called, we'd need to be concerned about potentially being blocked for the duration of the offline mode. 

One thing that you have to solve is what to return in the first call (before the first central reacheable event is received). We can assume that nothing _could_ be sent until this point, but if some Sensor component is already processing events and wants to send a message down the pipeline, that shouldn't be a problem. 

With that in mind, one idea would be to create a first context which is returned both in the first offline mode and when going online. Something like:

```mermaid
flowchart LR
    A[Init]
    C[Online]
    B[Offline]
    A --> aa[Create Initial Context]
    aa-->B
    B -->|CentralReachable|TryCreate{Is starting?}
    TryCreate -->|Yes|C
    TryCreate --> |No|Create[Swap context]
    Create -->C
    C -->|OfflineMode|Cancel[Cancel context]
    Cancel -->B
```

It's not an elegant solution, as it needs a special case for the first context. If you can come up with a solution that doesn't require special cases, even better.

Reply:
You are right. I initially went with the blocking behavior to avoid SensorComponents getting `nil` or incorrect `context` but I realized too that that would create a problem with components getting blocked in the middle of an execution (if we go offline) and on reconnect wrongly getting the new context.